lib/ramble/ramble/util/spec_utils.py

```diff
@@ -6,34 +6,120 @@
 
-    if "application_name" in all_keys:
-        all_keys.remove("application_name")
 
-    if "spec_type" in all_keys:
-        all_keys.remove("spec_type")
+def specs_conflict(new, existing, prefix="", skip_conflicting_when=False):
+    # Short circuit check if when clauses conflict
+    # (so specs should not be applied at the same time)
+    # Used for printing conflicting software specs.
+    if skip_conflicting_when:
+        new_when = set(new["when"]) if "when" in new else None
+        existing_when = set(existing["when"]) if "when" in existing else None
 
-    if "package_manager" in all_keys:
-        all_keys.remove("package_manager")
+        if new_when != existing_when:
+            return False
 
-    if "when" in all_keys:
-        all_keys.remove("when")
+    prefixed_keys = {}
+    for key in new.keys():
+        if new[key] is not None:
+            prefixed_keys[key] = f"{prefix}{key}"
 
-    for key in all_keys:
-        if key not in spec1:
-            return False
-        if key not in spec2:
-            return False
-        if spec1[key] != spec2[key]:
+    for in_key, out_key in prefixed_keys.items():
+        if out_key in existing and new[in_key] != existing[out_key]:
+            return True
+    return False
+
+
+class SoftwareSpec:
+
+    def __init__(
+        self,
+        name: str,
+        pkg_spec: str,
+        prefix: str = "",
+        compiler: str = None,
+        compiler_spec: str = None,
+        when: List[str] = None,
+    ):
+        self.name = name
+        self.pkg_spec = pkg_spec
+        self.prefix = prefix
+        self.compiler = compiler
+        self.compiler_spec = compiler_spec
+        self.when = when.copy()
+
+    def to_dict(self, prefix: str = None):
+        prefix_base = prefix if prefix is not None else self.prefix
+        prefix_str = f"{prefix_base}_" if prefix_base else ""
+
+        attrs = ["pkg_spec", "compiler", "compiler_spec"]
+
+        output = {}
+        for attr in attrs:
+            val = getattr(self, attr, None)
+            if val is not None:
+                output[f"{prefix_str}{attr}"] = val
+        return output
+
+    def config_opts(self):
+        self_dict = self.to_dict()
+        for key, val in self_dict.items():
+            yield f"software:packages:{self.name}:{key}:{val}"
+
+    def as_str(self, indent=0):
+        base_indent = " " * indent
+        indentation = " " * (indent + 4)
+        self_dict = self.to_dict()
+        color_name = ramble.util.colors.section_title(self.name)
+        output = f"{base_indent}{color_name}:\n"
+        for key, val in self_dict.items():
+            color_key = ramble.util.colors.nested_1(key)
+            escaped_val = val.replace("@", "@@")
+            output += f"{indentation}{color_key}: {escaped_val}\n"
+        return output
+
+    def __str__(self):
+        self_dict = self.to_dict(prefix="")
+        self_dict["prefix"] = self.prefix
+        return str(self_dict)
+
+    def copy(self):
```

Comment:
Is this any different from just calling something like `deepcopy`?

Reply:
Nope, it's not. I pushed an update that swaps it.